src/main/scala/org/neo4j/spark/writer/Neo4jDataWriter.scala

```diff
@@ -0,0 +1,115 @@
+package org.neo4j.spark.writer
+
+import java.util
+import java.util.Collections
+import java.util.concurrent.CountDownLatch
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SaveMode
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.sources.v2.writer.{DataWriter, WriterCommitMessage}
+import org.apache.spark.sql.types.StructType
+import org.neo4j.driver.exceptions.{Neo4jException, ServiceUnavailableException, SessionExpiredException}
+import org.neo4j.driver.{Session, Transaction, Values}
+import org.neo4j.spark.service.{MappingService, Neo4jQueryService, Neo4jQueryWriteStrategy}
+import org.neo4j.spark.util.Neo4jUtil._
+import org.neo4j.spark.util.Neo4jImplicits._
+import org.neo4j.spark.util.Neo4jUtil
+import org.neo4j.spark.{DriverCache, Neo4jOptions}
+
+class Neo4jDataWriter(jobId: String,
+                      partitionId: Int,
+                      structType: StructType,
+                      saveMode: SaveMode,
+                      options: Neo4jOptions) extends DataWriter[InternalRow] with Logging {
+
+  private val driverCache: DriverCache = new DriverCache(options.connection, jobId)
+
+  private var transaction: Transaction = _
+  private var session: Session = _
+
+  private val mappingService = new MappingService(options)
+
+  private val batch: util.List[java.util.Map[String, Object]] = new util.ArrayList[util.Map[String, Object]]()
+
+  private val retries = new CountDownLatch(options.transactionMetadata.retries)
+
+  val query: String = new Neo4jQueryService(options, new Neo4jQueryWriteStrategy(saveMode)).createQuery()
+
+  override def write(record: InternalRow): Unit = {
+    batch.add(mappingService.toParameter(record, structType))
+    if (batch.size() == options.nodeMetadata.batchSize) {
+      writeBatch
+    }
+  }
+
+  private def writeBatch(): Unit = {
+    if (session == null || !session.isOpen) {
+      session = driverCache.getOrCreate().session(options.session.toNeo4jSession)
+    }
+    if (transaction == null || !transaction.isOpen) {
+      transaction = session.beginTransaction()
```

Comment:
beginWriteTransaction?    Explicit

Reply:
we explicit it in the session with `AccesMode.WRITE`